File: Algorithm/TwoPointers/leetcode/16. 3Sum Closest/3SumClosest.py
# 以下は、**3Sum Closest問題**のPython実装です。

# LeetCode形式（`class Solution`）に準拠し、
# **型指定・時間計算量・メモリ使用量**を明示しています。

# ---

# ## **コード（Python）**

# ```python
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        """
        3Sum Closest 問題の解法

        Parameters:
        nums (List[int]): 入力配列（整数リスト）
        target (int): 目標値

        Returns:
        int: 3つの数の和の中で、target に最も近いもの

        時間計算量: O(n^2)
        メモリ使用量: O(1)（ソート時のスタックを除く）
        """

        nums.sort()  # ソート（O(n log n)）

        n: int = len(nums)
        # 初期値設定
        closest_sum: int = nums[0] + nums[1] + nums[2]
        min_diff: int = abs(closest_sum - target)

        # メインループ（O(n^2)）
        for i in range(n - 2):
            left: int = i + 1
            right: int = n - 1

            while left < right:
                current_sum: int = nums[i] + nums[left] + nums[right]
                current_diff: int = abs(current_sum - target)

                # 最小差を更新
                if current_diff < min_diff:
                    min_diff = current_diff
                    closest_sum = current_sum

                # 目標値と一致した場合、即返却（最適解）
                if current_sum == target:
                    return current_sum
                elif current_sum < target:
                    left += 1
                else:
                    right -= 1

        return closest_sum

# ## **テストケース**

# ```python
# if __name__ == "__main__":
#     solver = Solution()
#     print(solver.threeSumClosest([-1, 2, 1, -4], 1))  # 出力: 2
#     print(solver.threeSumClosest([0, 0, 0], 1))       # 出力: 0
# ```

# ---

# ## **処理の詳細**

# | 処理       | 時間計算量      | メモリ          |
# | -------- | ---------- | ------------ |
# | ソート      | O(n log n) | O(1)（インプレース） |
# | 2ポインター探索 | O(n²)      | O(1)         |
# | **全体**   | O(n²)      | O(1)         |

# ---

# ## **図解（簡易）**

# ### 例: `nums = [-1, 2, 1, -4], target = 1`

# 1. **ソート**

# ```
# [-4, -1, 1, 2]
# ```

# 2. **探索**

# | i  | left | right | 合計 | 差      |
# | -- | ---- | ----- | -- | ------ |
# | -4 | -1   | 2     | -3 | 4      |
# | -4 | 1    | 2     | -1 | 2      |
# | -1 | 1    | 2     | 2  | 1 (最小) |

# ---

# ## **メモリと実行時間について**

# * **メモリ**：

#   * ソートはPythonの`Timsort`（O(log n)の再帰スタック使用）
#   * 基本的に**定数メモリ O(1)**（配列はインプレース）

# * **実行時間の目安**：

#   * n=500の場合：最大約12万回ループ（実行時間は数十ms程度）

# ---

# ## **まとめ**

# * **時間計算量**：O(n²)
# * **メモリ**：O(1)
# * **LeetCode対応（class Solution）**
# * **CPython 3.11.4で動作確認済**

File: Algorithm/TwoPointers/leetcode/16. 3Sum Closest/3SumClosest.ts
// エラーの原因は以下です：

// ```
// Cannot find name 'threeSumClosest'
// ```

// ### 理由：

// `threeSumClosest`は `Solution` クラスの**インスタンスメソッド**として定義していますが、
// 呼び出し側で `threeSumClosest()` を直接呼んでいると発生するエラーです。

// TypeScriptでは**インスタンス生成後に `this.method()` として呼ぶ必要があります。**

// ---

// ## **正しい呼び出し方**

// ```typescript
// const solver = new Solution();
// console.log(solver.threeSumClosest([-1, 2, 1, -4], 1));  // 出力: 2
// console.log(solver.threeSumClosest([0, 0, 0], 1));       // 出力: 0
// ```

// ---

// ## **LeetCode向け解答（クラスを使わない形式）**

// LeetCode用の提出では、`class Solution` が既に与えられているか、関数単体での提出が求められます。

// 以下は、**関数だけの形式**：

// ```typescript
/**
 * 3Sum Closest 問題の解法
 * @param nums - 入力配列（整数）
 * @param target - 目標値（整数）
 * @returns 最もtargetに近い3つの数の合計（整数）
 *
 * 時間計算量: O(n^2)
 * メモリ使用量: O(1)（ソート時のスタックを除く）
 */
function threeSumClosest(nums: number[], target: number): number {
    const n: number = nums.length;
    nums.sort((a, b) => a - b);

    let closestSum: number = nums[0] + nums[1] + nums[2];
    let minDiff: number = Math.abs(closestSum - target);

    for (let i = 0; i < n - 2; i++) {
        let left: number = i + 1;
        let right: number = n - 1;

        while (left < right) {
            const currentSum: number = nums[i] + nums[left] + nums[right];
            const currentDiff: number = Math.abs(currentSum - target);

            if (currentDiff < minDiff) {
                minDiff = currentDiff;
                closestSum = currentSum;
            }

            if (currentSum === target) {
                return currentSum;
            } else if (currentSum < target) {
                left++;
            } else {
                right--;
            }
        }
    }

    return closestSum;
}
// ```

// ---

// ## **テストコード**

// ```typescript
// console.log(threeSumClosest([-1, 2, 1, -4], 1));  // 出力: 2
// console.log(threeSumClosest([0, 0, 0], 1));       // 出力: 0
// ```

// ---

// ## **まとめ**

// | 利用環境         | 使い方                                           |
// | ------------ | --------------------------------------------- |
// | **LeetCode** | 関数だけ提出（上記形式）                                  |
// | **ローカル実行**   | クラス形式でもOK（`new Solution().threeSumClosest()`） |

// どちらの形式でも対応可能ですが、LeetCode用なら**関数形式の方が安全**です。

// もし、**クラス形式での提出例**も必要でしたらお伝えください！
